xclim/core/calendar.py

```diff
@@ -767,175 +764,88 @@ def resample_doy(doy: xr.DataArray, arr: xr.DataArray | xr.Dataset) -> xr.DataAr
+    -----
+    xclim assumes that indexes for greater-than-day frequencies are "floored" down to a daily resolution.
+    For example, the coordinate "2000-01-31 00:00:00" with a "M" frequency is assumed to mean a period
+    going from "2000-01-01 00:00:00" to "2000-01-31 23:59:59.999999".
 
-def time_bnds(group, freq: str) -> Sequence[tuple[cftime.datetime, cftime.datetime]]:
+    Similarly, it assumes that daily and finer frequencies yield indexes pointing to the period's start.
+    So "2000-01-31 00:00:00" with a "3H" frequency, means a period going from "2000-01-31 00:00:00" to
+    "2000-01-31 02:59:59.999999".
     """
-    Find the time bounds for a pseudo-period index.
-
-    As we are using datetime indices to stand in for period indices, assumptions regarding the period
-    are made based on the given freq. IMPORTANT NOTE: this function cannot be used on greater-than-day freq
-    that start at the beginning of a month, e.g. 'MS', 'QS', 'AS' -- this mirrors pandas behavior.
+    if isinstance(time, (xr.DataArray, xr.Dataset)):
+        time = time.indexes[time.name]
+    elif isinstance(time, (DataArrayResample, DatasetResample)):
+        time = time._full_index
 
-    Parameters
-    ----------
-    group : CFTimeIndex or DataArrayResample
-        Object which contains CFTimeIndex as a proxy representation for
-        CFPeriodIndex
-    freq : str
-        String specifying the frequency/offset such as 'MS', '2D', or '3T'
+    if freq is None and hasattr(time, "freq"):
+        freq = time.freq
+    if freq is None:
+        freq = xr.infer_freq(time)
+    elif hasattr(freq, "freqstr"):
+        # When freq is a Offset
+        freq = freq.freqstr
 
-    Returns
-    -------
-    Sequence[(cftime.datetime, cftime.datetime)]
-        The start and end times of the period inferred from datetime and freq.
+    freq_is_start = parse_offset(freq)[2]
 
-    Examples
-    --------
-    >>> from xarray import cftime_range
-    >>> from xclim.core.calendar import time_bnds
-    >>> index = cftime_range(
-    ...     start="2000-01-01", periods=3, freq="2QS", calendar="360_day"
-    ... )
-    >>> out = time_bnds(index, "2Q")
-    >>> for bnds in out:
-    ...     print(
-    ...         bnds[0].strftime("%Y-%m-%dT%H:%M:%S"),
-    ...         " -",
-    ...         bnds[1].strftime("%Y-%m-%dT%H:%M:%S"),
-    ...     )
-    ...
-    2000-01-01T00:00:00  - 2000-03-30T23:59:59
-    2000-07-01T00:00:00  - 2000-09-30T23:59:59
-    2001-01-01T00:00:00  - 2001-03-30T23:59:59
-    """
-    if isinstance(group, CFTimeIndex):
-        cfindex = group
-    elif isinstance(group, DataArrayResample):
-        if isinstance(group._full_index, CFTimeIndex):
-            cfindex = group._full_index
+    if isinstance(time, xr.CFTimeIndex):
+        period = xr.coding.cftime_offsets.to_offset(freq)
+        is_on = period.onOffset(time[0])
+        eps = pd.Timedelta(precision or "1U").to_pytimedelta()
+        day = pd.Timedelta("1D").to_pytimedelta()
```

Comment:
Maybe use `floor` on `time` here to remove the hours, minutes and seconds ?

Reply:
It was more complicated than that because cftime objects do not implement neither `floor` nor `normalize`... I used `replace` instead.